### PR TITLE
blockchain: working on concurrency issues

### PR DESCRIPTION
**Pull request by ebuchman**
Working on https://github.com/tendermint/tendermint/issues/1629 but this still doesn't seem to cut it. 

This reactor needs some work to make it easier to understand and debug ...

### CODE REVIEW

File: blockchain/pool.go [hunk 1]
```diff
@@ -556,17 +567,8 @@ OUTER_LOOP:
...
-			case <-bpr.pool.Quit():
-				bpr.Stop()
-				return
-			case <-bpr.Quit():
-				return
-			case <-bpr.redoCh:
-				bpr.reset()
-				continue OUTER_LOOP
-			}
+			bpr.Stop()
```
Comment (jaekwon) [inline blockchain/pool.go]: Can't just stop here, <-bpr.redoCh can happen if the block received was bad (e.g. byzantine peer).

Comment (jaekwon) [reply blockchain/pool.go]: Thus the comment, "now see if it's good.".  brp doesn't stop until it gets a good block.

File: blockchain/pool.go [hunk 2]
```diff
@@ -133,7 +133,7 @@ func (pool *BlockPool) removeTimedoutPeers() {
 			// curRate can be 0 on start
 			if curRate != 0 && curRate < minRecvRate {
 				err := errors.New("peer is not sending us data fast enough")
-				pool.sendError(err, peer.id)
+				go pool.sendError(err, peer.id)
```
Comment (jaekwon) [inline blockchain/pool.go]: We shouldn't be creating go-routines like this if we can help it.

In this case, it shouldn't be necessary because the errorsCh should be draining just fine.  So I'm against this modification which only serves to mask bugs.

File: blockchain/pool.go [hunk 3]
```diff
@@ -232,21 +232,24 @@ func (pool *BlockPool) AddBlock(peerID p2p.ID, block *types.Block, blockSize int
...
-		pool.Logger.Info("peer sent us a block we didn't expect", "peer", peerID, "curHeight", pool.height, "blockHeight", block.Height)
-		diff := pool.height - block.Height
+		pool.Logger.Info("peer sent us a block we didn't expect", "peer", peerID, "curHeight", poolHeight, "blockHeight", block.Height)
+		diff := poolHeight - block.Height
 		if diff < 0 {
 			diff *= -1
 		}
 		if diff > maxDiffBetweenCurrentAndReceivedBlockHeight {
-			pool.sendError(errors.New("peer sent us a block we didn't expect with a height too far ahead/behind"), peerID)
+			go pool.sendError(errors.New("peer sent us a block we didn't expect with a height too far ahead/behind"), peerID)
```
Comment (jaekwon) [inline blockchain/pool.go]: ditto

File: blockchain/pool.go [hunk 4]
```diff
@@ -292,7 +295,8 @@ func (pool *BlockPool) removePeer(peerID p2p.ID) {
 	for _, requester := range pool.requesters {
 		if requester.getPeerID() == peerID {
 			if requester.getBlock() != nil {
-				pool.numPending++
+				continue
```
Comment (jaekwon) [inline blockchain/pool.go]: This is wrong?

File: blockchain/pool.go [hunk 5]
```diff
@@ -292,7 +295,8 @@ func (pool *BlockPool) removePeer(peerID p2p.ID) {
...
+				//				pool.numPending++
 			}
 			go requester.redo() // pick another peer and ...
```
Comment (jaekwon) [inline blockchain/pool.go]: ditto

Comment (jaekwon) [reply blockchain/pool.go]: This is from me I guess.

File: evidence/reactor.go [hunk 6]
```diff
@@ -118,21 +119,48 @@ func (evR *EvidenceReactor) broadcastRoutine() {
...
-			// TODO: Broadcast runs asynchronously, so this should wait on the successChan
-			// in another routine before marking to be proper.
+			// TODO: the broadcast here is just doing TrySend.
+			// We should make sure the send succeeds before marking broadcasted.
 			evR.evpool.evidenceStore.MarkEvidenceAsBroadcasted(evidence)
 		case <-ticker.C:
 			// broadcast all pending evidence
 			msg := &EvidenceListMessage{evR.evpool.PendingEvidence()}
-			evR.Switch.Broadcast(EvidenceChannel, cdc.MustMarshalBinaryBare(msg))
+			evR.broadcastEvidenceListMsg(msg)
```
Comment (melekes) [inline evidence/reactor.go]: should we mark evidence as broadcasted here too?

File: blockchain/pool.go (hunk 4 above)
Comment (ebuchman) [reply blockchain/pool.go]: Is it ? We already have the block, so why should we increase numPending? I think this is why I was seeing ~8000 numPending.

AFAIK, numPending should track how many requests are open but still waiting for a block. Here we already have the block - so why would this be wrong?

File: blockchain/pool.go (hunk 2 above)
Comment (ebuchman) [reply blockchain/pool.go]: Yeh I'm not thrilled with it but was trying to check if this kind of thing would help since it seemed like theres a deadlock somewhere and this is called within a locked function and sends on a channel. Note we already were doing this for the `redo()` below ...

File: blockchain/pool.go (hunk 1 above)
Comment (ebuchman) [reply blockchain/pool.go]: oh right. was hard to follow. we should make the comment clearer

File: blockchain/pool.go (hunk 4 above)
Comment (jaekwon) [reply blockchain/pool.go]: Because we're making the requester make the request again.
To be fair, this isn't the right place to do it, as there is a race condition.
I think maybe the requester should numPending... anyways, it should happen.

Comment (ebuchman) [reply blockchain/pool.go]: why make the request again if we already got the block ? because presumably its a bad peer so we dont want to try its block ?

I think the reason numPending shoots then is because we dont clear the block here. so when we get the block again from a new peer and call `setBlock`, it returns false and we dont decrement.

**Comment by ebuchman** [comment]
closing for https://github.com/tendermint/tendermint/pull/1633
